clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java

```diff
@@ -97,14 +121,47 @@ public void unsubscribe() {
     }
 
     @Override
-    public synchronized ConsumerRecords<K, V> poll(long timeout) {
+    public ConsumerRecords<K, V> poll(long timeout) {
         ensureNotClosed();
+
+        CountDownLatch pollLatchCopy = pollLatch.get();
+        if (pollLatchCopy != null) {
+            pollLatch.set(null);
+            pollLatchCopy.countDown();
+            synchronized (pollLatchCopy) {
+                // Will block until caller of waitUntilPollThen() finishes their callback.
+            }
+        }
+
+        if (wakeup.get()) {
+            wakeup.set(false);
+            throw new ConsumerWakeupException();
+        }
+
+        if (exception != null) {
+            RuntimeException exception = this.exception;
+            this.exception = null;
+            throw exception;
+        }
+
+        // Handle seeks that need to wait for a poll() call to be processed
+        for (TopicPartition tp : subscriptions.missingFetchPositions()) {
+            if (subscriptions.isOffsetResetNeeded(tp)) {
+                resetOffsetPosition(tp);
+            } else if (subscriptions.committed(tp) == null) {
```

Comment:
`seek()` should set `hasValidPosition` to true for the topic partition in `subscriptions`. So that topic partition should not be included in `missingFetchPositions`.

The code in KafkaConsumer looks like this:
```
if (!subscriptions.hasAllFetchPositions())
            updateFetchPositions(this.subscriptions.missingFetchPositions());
```
But I'm really unclear why we check `hasAllFetchPositions`. I guess just to avoid allocating the extra collection? Otherwise, I basically inlined the code for updating fetch position. I think the thing that is currently missing is the `coordinator.refreshCommittedOffsetsIsNeeded()` call, but I hadn't any use for it in this code since I'm not using consumer groups or offset commits.